### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by guypilo** [websocket-client/websocket-client#255]
Hey

Im trying to use this websocket package (version 0.37.0) and running few tests,
During one of my test I figure out that the memory usage is raising to 330mb when sending 2mb file and not drop down back until Im running gc.collect() and then it still leak (or something) about 3mb.
It can reproduce easily like that:

```
import websocket
import ssl
x = websocket.WebSocket(sslopt={"cert_reqs": ssl.CERT_NONE, "check_hostname": False})
x.connect("wss://HOSTNAME/api")
y = file("FILE_PATH","rb").read()
x.send(y)
```

where FILE_PATH is path for file of 2mb size.

Did anyone have the same issue?

Thanks

**Comment by larytet** [comment]
Hi,

I have traced the memory consumption to two issues
*   pip install wsaccel was missing
*   call to trace in the send_frame() 
```
diff --git a/_core.py b/_core.py
index 13f7593..4eab85a 100644
--- a/_core.py
+++ b/_core.py
@@ -257,7 +257,7 @@ class WebSocket(object):
             frame.get_mask_key = self.get_mask_key
         data = frame.format()
         length = len(data)
-        trace("send: " + repr(data))
+        #trace("send: " + repr(data))
 
         with self.lock:
             while data:

```
These two things help to reduce the memory footprint by more than 80%

**Comment by dacotagh** [comment]
The problem is 'range' is using instead of 'xrange'.

_abnf.py:    def _mask(_m, _d):

        for i in range(len(_d)):
            _d[i] ^= _m[i % 4]

For 2Mb data range creates list of 2 000 000 objects!

Please use xrange here, which is object with the same behavior and no memory consumption.

Of course, using wsaccel is hide this problem, but it is ridiculous.

**Comment by sawt00th** [comment]
@dacotagh xrange not possible for Python 3
I found another place in _abnf.py file : 
`#339 shortage = bufsize - sum(len(x) for x in self.recv_buffer)`
can will replace to : 
`shortage = bufsize - sum(map(len, self.recv_buffer))`
and uses only 4 call function instead of len of  self.recv_buffer * 2
![optima](https://cloud.githubusercontent.com/assets/14891592/19985165/10ddeaa0-a21a-11e6-89cb-ab80be7c1849.png)

**Comment by CL545740896** [comment]
> Hi,
> 
> I have traced the memory consumption to two issues
> 
> * pip install wsaccel was missing
> * call to trace in the send_frame()
> 
> ```
> diff --git a/_core.py b/_core.py
> index 13f7593..4eab85a 100644
> --- a/_core.py
> +++ b/_core.py
> @@ -257,7 +257,7 @@ class WebSocket(object):
>              frame.get_mask_key = self.get_mask_key
>          data = frame.format()
>          length = len(data)
> -        trace("send: " + repr(data))
> +        #trace("send: " + repr(data))
> 
>          with self.lock:
>              while data:
> ```
> These two things help to reduce the memory footprint by more than 80%
 do it?
        if self.get_mask_key:
            frame.get_mask_key = self.get_mask_key
        data = frame.format()
        length = len(data)
        trace("send: " + repr(data))

        with self.lock:
            while data:
                l = self._send(data)
                data = data[l:]

        return length

**Comment by engn33r** [comment]
Thanks for the comments everyone - I realize this is an old issue. Two items were fixed here that should help the memory issue:
1. The fix for issue #433 should have resolved the main issue here, which is the old `_mask()` function used `for i in range`, which "is ridiculous" as @dacotagh mentioned. I implemented the improvement in commit  287970e and the new `_mask()` function should be about 4 times faster for this operation with improved memory usage.
2. @sawt00th pointed out another area for improvement, using `map()` which is more performance optimized than a `for` loop. I implemented this with commit 60e4711 and this operation should be about 2 times faster based on my tests.
